lib/wibox/widget/template.lua

```diff
@@ -0,0 +1,90 @@
+---------------------------------------------------------------------------
+-- An abstract widget that handles a preset of concrete widget.
+--
+-- The `wibox.widget.template` widget is an abstraction layer that contains a
+-- concrete widget definition. The template widget can be used to build widgets
+-- that the user can customize at their will, thanks to the template mechanism.
+--
+-- @author Aire-One
+-- @copyright 2021 Aire-One <aireone@aireone.xyz>
+--
+-- @classmod wibox.widget.template
+-- @supermodule wibox.widget.base
+---------------------------------------------------------------------------
+
+local wbase = require("wibox.widget.base")
+local gtable = require("gears.table")
+local gtimer = require("gears.timer")
+
+local default_update_callback = function()
+    return nil
+end
```

Comment:
It seems quite inefficient to execute the whole of `:update()` just to run this no-op if the user didn't provide a `update_callback`.
I think it would make more sense to allow `self.update_callback == nil` and add something like this to `:update()`:
```lua
if type(self.update_callback) ~= "function" then
	return
end
```